src/header.iced

```diff
@@ -0,0 +1,90 @@
+{prng} = require('crypto')
+{createHash} = require('crypto')
+msgpack = require('purepack')
+crypto = require('keybase-nacl')
+nonce = require('./nonce.iced')
+
+compute_mac_key = (encryptor, header_hash, pubkey) ->
+  zero_bytes = Buffer.alloc(32)
+  mac_box = encryptor.encrypt({plaintext : zero_bytes, nonce : nonce.nonceForMACKeyBox(header_hash), pubkey})
+  return mac_box.slice(-32)
+
+exports.generate_encryption_header_packet = (encryptor, recipients) ->
+  mode = 0
+  header_list = []
+  header_list.push('saltpack')
+  header_list.push([1, 0])
+  header_list.push(mode)
+
+  payload_encryptor = crypto.alloc({force_js : true})
+  payload_key = prng(32)
+  payload_encryptor.secretKey = payload_key
+  ephemeral_encryptor = crypto.alloc({force_js : true})
+  ephemeral_encryptor.genBoxPair()
+  header_list.push(ephemeral_encryptor.publicKey)
+
+  sender_sbox = payload_encryptor.secretbox({plaintext : encryptor.publicKey, nonce : nonce.nonceForSenderKeySecretBox()})
+  header_list.push(sender_sbox)
+
+  recipients_list = []
+  for rec_pubkey in recipients
+    rec_pair = []
+    rec_pair.push(rec_pubkey)
+    rec_payload = ephemeral_encryptor.encrypt({plaintext : payload_key, nonce : nonce.nonceForPayloadKeyBox(), pubkey : rec_pubkey})
+    rec_pair.push(rec_payload)
+    recipients_list.push(rec_pair)
+  header_list.push(recipients_list)
+
+  crypto_hash = createHash('sha512')
+  header_intermediate = msgpack.pack(header_list)
+  crypto_hash.update(header_intermediate)
+  header_hash = crypto_hash.digest()
+  header_packet = msgpack.pack(header_intermediate)
+
+  mac_keys = []
+  for rec_pubkey in recipients
+    mac_keys.push(compute_mac_key(encryptor, header_hash, rec_pubkey))
+
+  return {header_list, header_hash, header_packet, mac_keys}
+
+exports.parse_encryption_header_packet = (decryptor, header_packet) ->
+  #unpack header
+  crypto_hash = createHash('sha512')
+  header_intermediate = msgpack.unpack(header_packet)
+  crypto_hash.update(header_intermediate)
+  header_hash = crypto_hash.digest()
+  header_list = msgpack.unpack(header_intermediate)
```

Comment:
After you've deserialized `header_list`, you can do a destructuring assignment to make it easier to pick out the different fields:

    [name, [major, minor], mode, secret, sender, recipeints] = header_list

I think that even does the correct (permissive) thing when there are additional fields beyond the 5th, for forward compatibility.